File: web/lib/oath-engine/stage-4-select.ts
// stage 4: pure-logic winner selection. mirrors pipeline/v2/stage_4_select.py

import 'server-only';
import rubricJson from './rubric.json';
import type { Candidate, Critique, SelectionResult, RubricScores } from './types';

const THRESHOLDS = rubricJson.thresholds;
const AXIS_NAMES = Object.keys(rubricJson.axes) as Array<keyof RubricScores>;

function passesThreshold(scores: RubricScores, overall: number): boolean {
  if (overall < THRESHOLDS.minimum_overall_score) return false;
  for (const name of AXIS_NAMES) {
    if ((scores[name] ?? 0) < THRESHOLDS.minimum_axis_score) return false;
  }
  return true;
}

export function selectWinner(candidates: Candidate[], critiques: Critique[]): SelectionResult {
  if (!candidates.length || candidates.length !== critiques.length) {
    throw new Error('candidates and critiques must be non-empty and same length');
  }

  const pairs = candidates.map((c, i) => ({ candidate: c, critique: critiques[i] }));
  const passing = pairs.filter((p) => passesThreshold(p.critique.scores, p.critique.overall_score));

  if (passing.length > 0) {
    const winnerPair = passing.reduce((best, cur) =>
      cur.critique.overall_score > best.critique.overall_score ? cur : best
    );
    return {
      winner: winnerPair.candidate,
      winner_critique: winnerPair.critique,
      passes_threshold: true,
      refinement_needed: false,
      selection_reason: `candidate ${winnerPair.candidate.candidate_id} passed thresholds with overall ${winnerPair.critique.overall_score}/10 (best of ${passing.length} passing)`,
    };
  }

  const winnerPair = pairs.reduce((best, cur) =>
    cur.critique.overall_score > best.critique.overall_score ? cur : best
  );
  return {
    winner: winnerPair.candidate,
    winner_critique: winnerPair.critique,
    passes_threshold: false,
    refinement_needed: true,
    selection_reason: `no candidate passed thresholds. picked candidate ${winnerPair.candidate.candidate_id} as closest-to-passing (overall ${winnerPair.critique.overall_score}/10, weakest axis: ${winnerPair.critique.weakest_axis})`,
  };
}

File: pipeline/v2/stage_4_select.py
"""stage 4 — candidate selection. pure logic, no claude call.

1. filter candidates that pass minimum thresholds
2. if any pass: pick highest overall_score
3. if none pass: pick closest-to-passing (highest overall), flag refinement_needed
"""

from rubric import passes_threshold


def select_winner(candidates: list, critiques: list) -> dict:
    if not candidates or not critiques or len(candidates) != len(critiques):
        raise ValueError("candidates and critiques must be non-empty and same length")

    # zip into (candidate, critique) tuples
    pairs = list(zip(candidates, critiques))

    # try threshold-passing candidates first
    passing = [
        (c, cr) for c, cr in pairs
        if passes_threshold(cr["scores"], cr["overall_score"])
    ]

    if passing:
        winner, winner_critique = max(passing, key=lambda p: p[1]["overall_score"])
        return {
            "winner": winner,
            "winner_critique": winner_critique,
            "passes_threshold": True,
            "refinement_needed": False,
            "selection_reason": (
                f"candidate {winner['candidate_id']} passed thresholds with overall "
                f"{winner_critique['overall_score']}/10 (best of {len(passing)} passing)"
            ),
        }

    # nothing passed — pick highest overall, flag refinement needed
    winner, winner_critique = max(pairs, key=lambda p: p[1]["overall_score"])
    return {
        "winner": winner,
        "winner_critique": winner_critique,
        "passes_threshold": False,
        "refinement_needed": True,
        "selection_reason": (
            f"no candidate passed thresholds. picked candidate {winner['candidate_id']} "
            f"as closest-to-passing (overall {winner_critique['overall_score']}/10, "
            f"weakest axis: {winner_critique['weakest_axis']})"
        ),
    }
